Installation tips using uv (#3503)
SpikeInterface/spikeinterface: installation_tips/README.md

## Installation tips

If you are not (yet) an expert in Python installations, a main difficulty is choosing the installation procedure.
The main ideas you need to know before starting:
 * python itself can be distributed and installed many many ways.

>  * python itself do not contain so many features for scientifique computing you need to install "packages".

Reviewer: ```suggestion
 * python itself does not contain so many features for scientific computing you need to install "packages".
```

>    Numpy, scipy, matplotlib, spikeinterface, ... are python packages that have a complicated dependency graph between then. "uv"

Reviewer: ```suggestion
   numpy, scipy, matplotlib, spikeinterface, ... are python packages that have a complicated dependency graph between then. "uv"
```

Reviewer: Agreed. Wasn't sure what he meant there. haha.

 * installing package  can be distributed and installed several ways (pip, conda, uv, mamba, ...)

>  * installing many packages at once is challenging (because of the depenency graph) so you need to do it in an "isolated environement"

Reviewer: ```suggestion
 * installing many packages at once is challenging (because of their dependency graphs) so you need to do it in an "isolated environement"
```

>    to not destroy any previous installation. You need to see an "environement" as a sub installtion in dedicated folder.

Reviewer: ```suggestion
   to not destroy any previous installation. You need to see an "environment" as a sub installation in a dedicated folder.
```

> Choosing the installator + a environement manager + a package installer is a nightmare for beginners.

Reviewer: ```suggestion
Choosing the installer + an environment manager + a package installer is a nightmare for beginners.
```

The main options are:

>   * use "anaconda" that do everything. The most popular but bad idea because : ultra slow and agressive licensing (not free anymore)

Reviewer: ```suggestion
  * use "anaconda", which does everything. The most popular but bad idea because : ultra slow and aggressive licensing (not always free anymore)
```

>   * use "anaconda" that do everything. The most popular but bad idea because : ultra slow and agressive licensing (not free anymore)

Reviewer 2: Maybe it is better to lead with the recommended one (i.e. uv first bullet point)?

On conda, is true in some cases but newer versions use `mamba` backend which is quite fast, and `conda-forge` does not have licensing issues. A problem is that HPC often come with the older slow versions and the default channel is anaconda's. 

No point going into too much detail but a suggestion to amend to say that that the default anaconda channel has agressive licensing and that it can be slow / is slower than uv. It is probably worth mentioning it can handle non-python dependencies while the other two can't, but this is not required for spikeinterface.

Author: I will try to modify a bit

Reviewer 3: conda+mamba via https://github.com/conda-forge/miniforge is my favorite, doesn't have licensing problems, and almost required when setting up gpu dependencies

  * use python from the system or python.org + venv + pip : good idea for linux users.

>   * use "uv" : a new, fast and simple. We recommand this for beginners on evry os.

Reviewer: ```suggestion
  * use "uv" : a new, fast and simple package manager. We recommend this for beginners on every os.
```

> Here we propose a steps by step recipe for beginers based on "uv".

Reviewer: ```suggestion
Here we propose a step by step recipe for beginers based on "uv".
```

> We used to propose here a solution based on anaconda. It is kept here for a while but we do not recommand it anymore.

Reviewer: ```suggestion
We used to recommend installing with anaconda. It will be kept here for a while but we do not recommend it anymore.
```

This environment will install:
 * spikeinterface full option
 * spikeinterface-gui
 * kilosort4

Kilosort, Ironclust and HDSort are MATLAB based and need to be installed from source.

### Quick installation

> 1. On macOS and Linux. Open a terminal and do

Reviewer 4: Two comments:
1) I think it would be better to refer them to the uv installation guide. If it changes, the our docs will be outdated for no reason and they have os dependent paths:

https://docs.astral.sh/uv/getting-started/installation/

2) Why are we creating new requirements? for the stabel if uv installs from pip it should suffice, for the rolling, uv can install directly from github and that would cover that. 

In both cases it would use the pyproject.toml.

Reviewer: I think this is a good point for these external tools!

Author: I think I do not want to overload the pyproject
The idea to have a working env for a beginner for a tutorial that contains spikeinterface but also external tools.
sigui sortingview ks.
The idea of having a entire and unqiue recipe and not refering to uv is to avoid a a benginer to understand evreything and compiling informations from several places.

Reviewer 4: I don't understand, you don't need to add anything to the pyproject. The point of the proposal is to have less files on the repo.

Author: Here the requirements is bug : PySide6, jupyter.
I do not see this requirements as optional dependencies of spikeinterface but more a minimal stack for a beginner.
what would be the labels in pyproject.toml for theses 2 requirements files ?

Reviewer 4: Fair enough, the right way would be to use the newly accepted [PEP 735](https://peps.python.org/pep-0735/) that covers just this but that is not available yet. 

I would suggest that it would be better for you (so you don't have to remember about dependencies in two places) if the requirements files only had the dependencies that are not in the pyprojec.toml there. Then you could do:

```bash
uv pip install spikeinterface
uv pip install `the_beginnger_required_stack.txt` 
```

But maybe you don't mind having these not working some time.

   `$ curl -LsSf https://astral.sh/uv/install.sh | sh`

> 1. On windows. Open a powershell and do

Reviewer: Hey Sam we recommend users use Command Prompt since our shell scripts for sorters only work in command prompt. It might be better if they have a command prompt version of instructions.

Author: could you test if the windows command also works in CMD ?
uv docs recommand powershell (I tested and it works)

   `powershell -ExecutionPolicy ByPass -c "irm https://astral.sh/uv/install.ps1 | iex"`
2. exit session and log again.
3. Download with right click + save this file corresponding in "Documents" folder:
    * [`requirements_stable.txt`](https://raw.githubusercontent.com/SpikeInterface/spikeinterface/main/installation_tips/requirements_stable.txt)
4. open terminal or powershell
5. `uv venv si_env --python 3.11`
6. `source `source si_env/bin/activate` (you should have `(si_env)` in your terminal)
7. `uv pip install -r Documents/requirements_stable.txt`

More detail on [uv here](https://github.com/astral-sh/uv).

## Installing before release

Some tools in the spikeinteface ecosystem are getting regular bug fixes (spikeinterface, spikeinterface-gui, probeinterface, python-neo, sortingview).
We are making releases 2 to 4 times a year. In between releases if you want to install from source you can use the `requirements_rolling.txt` file to create the environment. This will install the packages of the ecosystem from source.

> This is a good way to test if patch fix your issue.

Reviewer: ```suggestion
This is a good way to test if a patch fixes your issue.
```

    You need to play with "community channels" to make it free again, which is too complicated for beginners.
    Do not go this way.
  * use python from the system or python.org + venv + pip : good idea for linux users.

Here we propose a step by step recipe for beginers based on **"uv"**.
We used to recommend installing with anaconda. It will be kept here for a while but we do not recommend it anymore.

This environment will install:
 * spikeinterface full option
 * spikeinterface-gui
 * kilosort4

Kilosort, Ironclust and HDSort are MATLAB based and need to be installed from source.

### Quick installation using "uv" (recommended)

1. On macOS and Linux. Open a terminal and do
   `$ curl -LsSf https://astral.sh/uv/install.sh | sh`
1. On windows. Open a powershell and do
   `powershell -ExecutionPolicy ByPass -c "irm https://astral.sh/uv/install.ps1 | iex"`
2. exit session and log again.
3. Download with right click and save this file corresponding in "Documents" folder:
    * [`requirements_stable.txt`](https://raw.githubusercontent.com/SpikeInterface/spikeinterface/main/installation_tips/requirements_stable.txt)
4. open terminal or powershell
5. `uv venv si_env --python 3.11`

> 6. `source si_env/bin/activate` (you should have `(si_env)` in your terminal)

Reviewer: ```suggestion
6. For Mac/Linux `source si_env/bin/activate` (you should have `(si_env)` in your terminal) or for Powershell `si_env\Scripts\activate` 
```

1. Download with right click + save the file [`check_your_install.py`](https://raw.githubusercontent.com/SpikeInterface/spikeinterface/main/installation_tips/check_your_install.py)
    and put it into the "Documents" folder
2. Open the CMD (Windows) or Terminal (Mac/Linux)

> 3. Acticate your is_env : `source si_env/bin/activate`

Reviewer: ```suggestion
3. Activate your si_env : `source si_env/bin/activate` (Max/Linux), `si_env\Scripts\activate` (CMD prompt)
```

  * opening the spikeinterface-gui
  * exporting to Phy

> ### Legacy installation using anaconda (not recomanded)

Reviewer: ```suggestion
### Legacy installation using anaconda (not recommended)
```
